zrml/court/src/lib.rs

```diff
@@ -1188,6 +1191,101 @@ mod pallet {
+                .last_join_at
+                .checked_div(&T::InflationPeriod::get())
+                .ok_or(Error::<T>::Unexpected(UnexpectedError::InflationPeriodIsZero))?;
+            if current_period_index != last_reward_period_index {
+                if current_period_index != last_join_period_index {
+                    return Ok((pool_item.stake, pool_item.last_join_at));
+                } else {
+                    return Ok((pool_item.pre_period_join_stake, pool_item.pre_period_join_at));
+                }
+            }
+
+            Ok((amount, now))
+        }
+
+        fn get_initial_joined_at(
+            prev_pool_item: Option<&CourtPoolItemOf<T>>,
+            now: BlockNumberFor<T>,
+        ) -> BlockNumberFor<T> {
+            match prev_pool_item {
+                Some(i) => i.joined_at,
+                None => now,
+            }
+        }
+
+        #[inline]
+        fn is_sorted_and_all_greater_than_lowest(
+            p: &CourtPoolOf<T>,
+            lowest_stake: BalanceOf<T>,
+        ) -> bool {
+            let mut sorted = p.clone();
+            sorted.sort_by_key(|pool_item| (pool_item.stake, pool_item.court_participant.clone()));
+            p.len() == sorted.len()
+                && p.iter().zip(sorted.iter()).all(|(a, b)| lowest_stake <= a.stake && a == b)
+        }
+
+        fn remove_weakest_if_full(
+            mut p: CourtPoolOf<T>,
+            amount: BalanceOf<T>,
+        ) -> Result<CourtPoolOf<T>, DispatchError> {
+            if p.is_full() {
+                let lowest_item = p.first();
+                let lowest_stake = lowest_item
+                    .map(|pool_item| pool_item.stake)
+                    .unwrap_or_else(<BalanceOf<T>>::zero);
+                debug_assert!(Self::is_sorted_and_all_greater_than_lowest(&p, lowest_stake));
+                ensure!(amount > lowest_stake, Error::<T>::AmountBelowLowestJuror);
+                p.remove(0);
+            }
+
+            Ok(p)
+        }
+
+        fn handle_existing_participant(
+            who: &T::AccountId,
+            amount: BalanceOf<T>,
+            mut pool: CourtPoolOf<T>,
+            prev_p_info: &CourtParticipantInfoOf<T>,
+        ) -> Result<(CourtPoolOf<T>, BalanceOf<T>, Option<CourtPoolItemOf<T>>), DispatchError>
+        {
+            ensure!(amount >= prev_p_info.stake, Error::<T>::AmountBelowLastJoin);
+
+            if let Some((index, pool_item)) = Self::get_pool_item(&pool, prev_p_info.stake, who)? {
+                let consumed_stake = pool_item.consumed_stake;
+                let prev_pool_item = Some(pool_item.clone());
+
+                pool.remove(index);
+
+                Ok((pool, consumed_stake, prev_pool_item))
+            } else {
+                let consumed_stake = prev_p_info.active_lock;
+
+                let pool = Self::remove_weakest_if_full(pool, amount)?;
+
+                Ok((pool, consumed_stake, None))
+            }
+        }
+
         fn do_join_court(
```

Comment:
Just realizing now that joining and extending stake are intermingled here which leads to a bit more case distinctions and also not intuitive interaction (as an existing Juror has to join twice with the full amount to extend existing stake). Low prio, but should be optimized at some point in the future.